From dd8b300927b198c0842c7bc0f2d064aeba41db79 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 13 Feb 2025 18:33:40 +0100
Subject: [PATCH 797/954] feat(stm32mp2): add sanity check in low-power context

Add sanity check in low-power context for consistency between BL2 and BL31:
- add version check (major change for backward compatibility issue)
- add size of provided data, used for CRC and checked for minor version
  change (BL2 receive at least the expected size)
- add CRC to protect the content of the low power context during
  Standby.

Using the ARMV8 intrinsic CRC32 for BL31 (tf_crc32.c and armv8-a+crc)
and the function provided by ZLIB for BL2.

STM32MP_CONTEXT_VERSION is introduced to avoid the backward compatibility
break with old BL2 binary when STM32MP_CONTEXT_VERSION=0.
But default OpenSTLinux use now the new pm context.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 2916515d76f3cc4ac8cd4f100691ee6d8f549af2 feat(stm32mp2): add sanity check in low-power context]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/423284
Change-Id: Ibbc3d2ac4acca8898db097a5ad42d292ac1d0fb8
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/437068
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/platform.mk        | 10 +++-
 plat/st/stm32mp2/stm32mp2_context.c | 80 +++++++++++++++++++++++++++--
 2 files changed, 86 insertions(+), 4 deletions(-)

diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index ab10ef83a6..e8b8af4d0c 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -83,6 +83,9 @@ else
 STM32_HEADER_VERSION_MINOR	:=	2
 endif
 
+# PM context version check between BL2 and BL31: break backward compatibility
+STM32MP_CONTEXT_VERSION		?=	1
+
 PKA_USE_NIST_P256		?=	0
 PKA_USE_BRAINPOOL_P256T1 	?=	0
 
@@ -196,6 +199,7 @@ $(eval $(call assert_booleans,\
 		PKA_USE_NIST_P256 \
 		STM32MP_CRYPTO_ROM_LIB \
 		STM32_SAES_CRYP2 \
+		STM32MP_CONTEXT_VERSION \
 		STM32MP_DDR_DUAL_AXI_PORT \
 		STM32MP_DDR_FIP_IO_STORAGE \
 		STM32MP_DDR3_TYPE \
@@ -233,6 +237,7 @@ $(eval $(call add_defines,\
 		STM32_SAES_VER \
 		STM32_SAES_CRYP2 \
 		STM32_TF_A_COPIES \
+		STM32MP_CONTEXT_VERSION \
 		STM32MP_CRYPTO_ROM_LIB \
 		STM32MP_DDR_DUAL_AXI_PORT \
 		STM32MP_DDR_FIP_IO_STORAGE \
@@ -373,7 +378,10 @@ BL31_SOURCES			+=	plat/st/stm32mp2/bl31_plat_setup.c			\
 					plat/st/stm32mp2/stm32mp2_topology.c
 
 ifeq ($(filter 1,${STM32MP_UART_PROGRAMMER} ${STM32MP_USB_PROGRAMMER}),)
-BL31_SOURCES			+=	plat/st/stm32mp2/stm32mp2_pm.c
+BL31_SOURCES			+=	plat/st/stm32mp2/stm32mp2_pm.c				\
+					common/tf_crc32.c
+
+BL31_CPPFLAGS += -march=armv8-a+crc
 endif
 
 ifeq ($(STM32MP_M33_TDCID),0)
diff --git a/plat/st/stm32mp2/stm32mp2_context.c b/plat/st/stm32mp2/stm32mp2_context.c
index 3edb1e5f61..4bdd6ae564 100644
--- a/plat/st/stm32mp2/stm32mp2_context.c
+++ b/plat/st/stm32mp2/stm32mp2_context.c
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+#include <common/tf_crc32.h>
 #include <drivers/clk.h>
 #include <lib/el3_runtime/context_mgmt.h>
 #include <lib/el3_runtime/cpu_data.h>
@@ -25,10 +26,22 @@
 #define DATA_SIZE		U(4)
 
 /* Magic used to indicated valid = ' ' 'M' 'P' '2' */
-#define CONTEXT_MAGIC			0x204D5032
+#define CONTEXT_MAGIC			U(0x204D5032)
+
+/* VERSION to check compatibility of pm backup data between bl2 and bl31 */
+#define ST_CTX_VER_MAJOR		1U
+#define ST_CTX_VER_MINOR		0U
+#define ST_CTX_VERSION			((ST_CTX_VER_MAJOR << 8U) | ST_CTX_VER_MINOR)
+#define ST_CTX_VER_GET_MAJOR(version)	(((version) & GENMASK_32(15, 8)) >> 8U)
+#define ST_CTX_VER_GET_MINOR(version)	((version) & GENMASK_32(7, 0))
 
 struct backup_data_s {
 	uint32_t magic;
+#if STM32MP_CONTEXT_VERSION
+	uint16_t version;
+	uint16_t size;
+	uint32_t crc_32;
+#endif
 	uint8_t enc_mkey[ENC_KEY_SIZE_IN_BYTES];
 	psci_power_state_t standby_pwr_state;
 	cpu_context_t saved_cpu_s_context[PLATFORM_CORE_COUNT];
@@ -42,6 +55,20 @@ struct backup_data_s {
 	uint8_t data[DATA_SIZE];
 };
 
+#if STM32MP_CONTEXT_VERSION
+static uint32_t stm32mp_pm_crc_check(struct backup_data_s *data)
+{
+	uint32_t saved_crc = data->crc_32;
+	uint32_t crc;
+
+	data->crc_32 = 0U;
+	crc =  tf_crc32(0U, (unsigned char *)data, data->size);
+	data->crc_32 = saved_crc;
+
+	return crc;
+}
+#endif
+
 void stm32mp_pm_save_enc_mkey_in_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
@@ -71,13 +98,45 @@ void stm32mp_pm_get_enc_mkey_from_context(uint8_t *data)
 bool stm32_pm_context_is_valid(void)
 {
 	struct backup_data_s *backup_data;
-	bool ret;
+	bool ret = true;
+#if STM32MP_CONTEXT_VERSION
+	uint32_t crc_32 = 0U;
+#endif
 
 	backup_data = (struct backup_data_s *)BACKUP_CTX_ADDR;
 
 	clk_enable(BACKUP_CTX_CLK);
 
-	ret = (backup_data->magic == CONTEXT_MAGIC);
+	if (backup_data->magic != CONTEXT_MAGIC) {
+		ret = false;
+	} else {
+#if STM32MP_CONTEXT_VERSION
+		if ((ST_CTX_VER_GET_MAJOR(backup_data->version) != ST_CTX_VER_MAJOR) ||
+		    (ST_CTX_VER_GET_MINOR(backup_data->version) < ST_CTX_VER_MINOR) ||
+		    (backup_data->size < sizeof(*backup_data))) {
+			ret = false;
+		} else {
+			crc_32 = stm32mp_pm_crc_check(backup_data);
+			if (crc_32 != backup_data->crc_32) {
+				ret = false;
+			}
+		}
+
+		if (!ret) {
+			ERROR("pm context invalid (crc=%x vs %x, size=%u vs %lu, ver=%x vs %x)\n",
+			      backup_data->crc_32, crc_32, backup_data->size, sizeof(*backup_data),
+			      backup_data->version, ST_CTX_VERSION);
+			backup_data->magic = 0U;
+		}
+#endif
+	}
+
+#if STM32MP_CONTEXT_VERSION
+	VERBOSE("%s(crc=%x vs %x, size=%u vs %lu, ver=%x vs %x) = %d\n", __func__,
+	        backup_data->crc_32, crc_32, backup_data->size, sizeof(*backup_data),
+	        backup_data->version, ST_CTX_VERSION, ret);
+#endif
+
 	clk_disable(BACKUP_CTX_CLK);
 
 	return ret;
@@ -127,6 +186,16 @@ void stm32_pm_context_save(const psci_power_state_t *state, const void *data, si
 	if ((data != NULL) && (size != 0U)) {
 		(void)memcpy(&backup_data->data, data, MIN(size, (size_t)DATA_SIZE));
 	}
+
+#if STM32MP_CONTEXT_VERSION
+	backup_data->version = ST_CTX_VERSION;
+	backup_data->size = (uint16_t)sizeof(*backup_data);
+	backup_data->crc_32 = stm32mp_pm_crc_check(backup_data);
+
+	VERBOSE("%s(crc=%x, size=%u, version=%x)\n", __func__,
+	        backup_data->crc_32, backup_data->size, backup_data->version);
+#endif
+
 	clk_disable(BACKUP_CTX_CLK);
 }
 
@@ -191,6 +260,11 @@ void stm32_pm_context_clear(void)
 	clk_enable(BACKUP_CTX_CLK);
 
 	backup_data->magic = 0U;
+#if STM32MP_CONTEXT_VERSION
+	backup_data->version = 0U;
+	backup_data->size = 0U;
+	backup_data->crc_32 = 0U;
+#endif
 
 	clk_disable(BACKUP_CTX_CLK);
 }
-- 
2.53.0

